tests/server/job/test_job.py

```diff
@@ -0,0 +1,322 @@
+        time.sleep(1)
+        job_runner_proc.kill()
+        job_runner_proc.wait()  # ensure the job runner process is killed.
+
+        # assert that job1 has done, job2 is running, and job3 is pending.
+        assert query_job(job1_id) == (JobStatus.DONE, 7)
+        assert query_job(job2_id) == (JobStatus.RUNNING, None)
+        assert query_job(job3_id) == (JobStatus.PENDING, None)
+
+        # restart the job runner, and verify it resumes unfinished jobs (job2 and job3)
+        _start_job_runner_for_test(1, False)
+        time.sleep(2.5)
+
+        # assert all jobs are done.
+        assert query_job(job1_id) == (JobStatus.DONE, 7)
+        assert query_job(job2_id) == (JobStatus.DONE, 11)
+        assert query_job(job3_id) == (JobStatus.DONE, 15)
+
+
+def test_job_resume_on_new_job_runner(monkeypatch, tmp_path):
+    db_tmp_path = tmp_path / "db"
+    db_tmp_path.mkdir()
+    runner1_tmp_path = tmp_path / "runner1"
+    runner1_tmp_path.mkdir()
+    runner2_tmp_path = tmp_path / "runner2"
+    runner2_tmp_path.mkdir()
+
+    backend_store_uri = f"sqlite:///{str(db_tmp_path / 'mlflow.db')}"
+
+    with _setup_job_runner(1, monkeypatch, runner1_tmp_path, backend_store_uri) as job_runner_proc:
+        job1_id = submit_job(basic_job_fun, {"x": 3, "y": 4, "sleep_secs": 0})
+        job2_id = submit_job(basic_job_fun, {"x": 5, "y": 6, "sleep_secs": 10})
+        job3_id = submit_job(basic_job_fun, {"x": 7, "y": 8, "sleep_secs": 0})
+        time.sleep(1)
+
+    # ensure the job runner process is killed.
+    job_runner_proc.wait()
+
+    with _setup_job_runner(1, monkeypatch, runner2_tmp_path, backend_store_uri):
+        # assert that job1 has done, job2 is running, and job3 is pending.
+        assert query_job(job1_id) == (JobStatus.DONE, 7)
+        assert query_job(job2_id) == (JobStatus.RUNNING, None)
+        assert query_job(job3_id) == (JobStatus.PENDING, None)
+        time.sleep(10)
+        # assert all jobs are done.
+        assert query_job(job1_id) == (JobStatus.DONE, 7)
+        assert query_job(job2_id) == (JobStatus.DONE, 11)
+        assert query_job(job3_id) == (JobStatus.DONE, 15)
+
+
+def test_job_queue_parallelism(monkeypatch, tmp_path):
+    # test job queue parallelism=2 and each job consumes 2 seconds.
+    with _setup_job_runner(2, monkeypatch, tmp_path):
+        job_ids = [submit_job(basic_job_fun, {"x": x, "y": 1, "sleep_secs": 2}) for x in range(4)]
+        time.sleep(2.5)
+
+        # assert that job1 and job2 are done, and job3 and job4 are running
+        assert query_job(job_ids[0]) == (JobStatus.DONE, 1)
+        assert query_job(job_ids[1]) == (JobStatus.DONE, 2)
+        assert query_job(job_ids[2]) == (JobStatus.RUNNING, None)
+        assert query_job(job_ids[3]) == (JobStatus.RUNNING, None)
+
+        time.sleep(2.5)
+        assert query_job(job_ids[2]) == (JobStatus.DONE, 3)
+        assert query_job(job_ids[3]) == (JobStatus.DONE, 4)
+
+
+def transient_err_fun(tmp_dir, succeed_on_nth_run):
```

Comment:
what is `tmp_dir` type? If it had a `Path` type hint, I could comment `Let's not use os.listdir` without even checking the caller. Type hints help reviewers.

Reply:
`tmp_dir` is `str type.  
we can't make it to be `Path` type because `Path` instance is not JSON-serializable.
(job function parasm needs to be JSON-serializable so that we can serialize them as JSON string and save JSON string in database backend)